Implement MSC3356 to support additional fields in the OpenID user info

jkanefendt opened the pull request:
Add option to include additional fields in the OpenID user info (currently the user's display name an effective power levels).
This should simplify room-based access control in third-party widgets. Currently, joined rooms and power levels can only be determined via Admin-API.

Signed-off-by: Johannes Kanefendt <johannes.kanefendt@krzn.de>

clokep commented:
I believe that this would need an MSC to accompany it since it is a specified endpoint? (In which case we wouldn't need to have any config options and could just include them, most likely.)

richvdh reviewed (changes requested):
> I believe that this would need an MSC to accompany it since it is a specified endpoint? (In which case we wouldn't need to have any config options and could just include them, most likely.)

Yeah, there's not much point in supporting extra fields for Synapses deployed on the open federation unless the spec is extended to document the additional fields.

That said, there's probably a usecase for custom fields in private federations - but in that case, I'd expect the extension to happen via a Synapse module (see https://matrix-org.github.io/synapse/latest/modules.html) rather than an eclectic mix of configuration options.

@jkanefendt: could describe your usecase here, so we can suggest how to proceed?

jkanefendt commented:
> > I believe that this would need an MSC to accompany it since it is a specified endpoint? (In which case we wouldn't need to have any config options and could just include them, most likely.)
> 
> Yeah, there's not much point in supporting extra fields for Synapses deployed on the open federation unless the spec is extended to document the additional fields.
> 
> That said, there's probably a usecase for custom fields in private federations - but in that case, I'd expect the extension to happen via a Synapse module (see https://matrix-org.github.io/synapse/latest/modules.html) rather than an eclectic mix of configuration options.
> 
> @jkanefendt: could describe your usecase here, so we can suggest how to proceed?

I finally found the time to create a MSC with a brief description and usecase - see https://github.com/matrix-org/matrix-doc/blob/4c415fb7bc2d991a4515820d8c4fda75e98ce94e/proposals/3356-add-openid-userinfo-fields.md

anoadragon453 commented:
@jkanefendt That [MSC](https://github.com/matrix-org/matrix-doc/pull/3356) will need some unstable prefixes defined (or the MSC otherwise needs to be merged) before we can implement it in Synapse. See [here](https://github.com/matrix-org/matrix-doc/blob/clokep/restricted-rooms/proposals/3083-restricted-rooms.md#unstable-prefix) for an example of defining unstable prefixes. You would need to add an unstable prefix to each of the new key names in the OpenID response.

It also appears that there are some unaddressed comments on the MSC, which addressing will help move the MSC forwards :slightly_smiling_face:

jkanefendt commented:
> @jkanefendt That [MSC](https://github.com/matrix-org/matrix-doc/pull/3356) will need some unstable prefixes defined (or the MSC otherwise needs to be merged) before we can implement it in Synapse. See [here](https://github.com/matrix-org/matrix-doc/blob/clokep/restricted-rooms/proposals/3083-restricted-rooms.md#unstable-prefix) for an example of defining unstable prefixes. You would need to add an unstable prefix to each of the new key names in the OpenID response.
> 
> It also appears that there are some unaddressed comments on the MSC, which addressing will help move the MSC forwards

Thanks for your feedback @anoadragon453 . I'm going to add the unstable prefixes ASAP.
In the meantime, if found some time for some security considerations regarding  [MSC3356](https://github.com/matrix-org/matrix-doc/pull/3356). My proposal for releasing fields on a per-token basis is implemented in https://github.com/jkanefendt/synapse/commit/92e663cdfa387c3f97b0c1f9c2a20cead0fe6dde.

anoadragon453 commented:
> In the meantime, i found some time for some security considerations regarding MSC3356.

I think that's an excellent addition. And since we're on the topic of implementing additional features as they come into the MSC, don't be afraid to break up the feature into multiple PRs, especially for self-contained changes like specifying which fields to return. It can help make review easier and get your code merged quicker.

Looking forward to the unstable prefixes. Do poke on the MSC PR if you need any further clarifications on how to add them.

jkanefendt commented:
> 
> Yeah, there's not much point in supporting extra fields for Synapses deployed on the open federation unless the spec is extended to document the additional fields.
> 
> That said, there's probably a usecase for custom fields in private federations - but in that case, I'd expect the extension to happen via a Synapse module (see https://matrix-org.github.io/synapse/latest/modules.html) rather than an eclectic mix of configuration options.
> 
> @jkanefendt: could describe your usecase here, so we can suggest how to proceed?

All these points have been covered in [MSC3356](https://github.com/jkanefendt/matrix-doc/blob/0aa92f65841deffb9c67bdfb1cd1b3014b1669ae/proposals/3356-add-openid-userinfo-fields.md) by now.

DMRobertson commented:
@jkanefendt  I think that number is already in use: https://github.com/matrix-org/matrix-doc/issues/3556. If you make a pull request to that repo from the branch you linked, you should be assigned a unique number. cc @matrix-org/spec-core-team .

anoadragon453 commented:
Hi @jkanefendt. Just want to double-check as the MSC is still in flux it seems. Would you be happy if this PR merged in its current form, or would you like to hold off until the MSC is further iterated upon?

jkanefendt commented:
> Hi @jkanefendt. Just want to double-check as the MSC is still somewhat in flux it seems. Would you be happy if this PR merged in its current form, or would you like to hold off until the MSC is further iterated upon?

Hi @anoadragon453. I don't expect any further changes for [MSC3356](https://github.com/jkanefendt/matrix-doc/blob/0aa92f65841deffb9c67bdfb1cd1b3014b1669ae/proposals/3356-add-openid-userinfo-fields.md) so i'd be happy to see this PR merged :+1:

reivilibre reviewed (changes requested):
This looks fairly fine as an implementation of the MSC, though as mentioned on the MSC I'm not sure about the privacy implications.

It should also be behind an experimental flag for now (this is standard procedure for MSCs that haven't become part of the stable spec).

anoadragon453 commented:
Hi there @jkanefendt. Are you able to continue work on this PR?

jkanefendt commented on synapse/rest/client/openid.py:
```diff
@@ -64,6 +65,12 @@ class IdTokenServlet(RestServlet):
+        item[1]
+        for item in OpenIdUserInfoField.__dict__.items()
+        if not item[0].startswith("_")
+    ]
```
See https://github.com/matrix-org/synapse/pull/10384/commits/ea34feefa2cfa2515b3040f86b68365863521265

jkanefendt commented on synapse/rest/client/openid.py:
```diff
@@ -78,14 +85,25 @@ async def on_POST(
+                if field not in self.USERINFO_FIELDS:
+                    raise SynapseError(
+                        400,
+                        "Unknown userinfo field '" + field + "'",
```
See https://github.com/matrix-org/synapse/pull/10384/commits/4f0ac8ed32a36e61eacc47877f9ed641c1604c81

jkanefendt commented on synapse/storage/databases/main/openid.py:
```diff
@@ -12,32 +12,41 @@
+        token: str,
+        ts_valid_until_ms: int,
+        user_id: str,
+        userinfo_fields: Optional[list],
```
https://github.com/matrix-org/synapse/pull/10384/commits/597dd31cb960c3313b1f6f474917ac3c9b3b6fc2

jkanefendt commented on synapse/rest/client/openid.py:
```diff
@@ -78,14 +85,25 @@ async def on_POST(
+        json = parse_json_object_from_request(request, allow_empty_body=True)
+
+        userinfo_fields = None
+        if "org.matrix.msc3356.userinfo_fields" in json:
```
See https://github.com/matrix-org/synapse/pull/10384/commits/b429c475de2496b00046580502f2065b0393bf42

jkanefendt commented on synapse/federation/transport/server/__init__.py:
```diff
@@ -283,7 +285,48 @@ async def on_GET(
+            )
+            room_power_levels = room_state.get((EventTypes.PowerLevels, ""))
+            if room_power_levels is not None:
+                stripped_content = copy.deepcopy(room_power_levels.content)
+                stripped_content["users"] = {
+                    k: v for k, v in stripped_content["users"].items() if k == user_id
+                }
+                power_levels[room.room_id] = stripped_content
```
At this point i followed @turt2live's suggestion to include the whole state event in verbatim, and to include only the current user's powerlevels in the `users` field. See https://github.com/matrix-org/matrix-spec-proposals/pull/3356#discussion_r755574795

Would it meet your concerns if we just froze the currently copied fields/hierarchy? I.e.:
```json
{
    "users": {
      "@jondoe:example.com": 100
    },
    "users_default": 0,
    "events": {
      "m.room.name": 50,
      "m.room.power_levels": 100,
      "m.room.history_visibility": 100,
      "m.room.canonical_alias": 50,
      "m.room.avatar": 50,
      "m.room.tombstone": 100,
      "m.room.server_acl": 100,
      "m.room.encryption": 100
    },
    "events_default": 0,
    "state_default": 50,
    "ban": 50,
    "kick": 50,
    "redact": 50,
    "invite": 0,
    "historical": 100
}
```

richvdh replied:
I don't think I have strong feelings on this. I think the current behaviour is probably fine. It would be nice if the MSC made the behaviour here explicit though.

richvdh commented on synapse/federation/transport/server/__init__.py:
```diff
@@ -286,7 +288,50 @@ async def on_GET(
+
+        power_levels = {}
+        for room in room_list:
+            room_state = await self.hs.get_state_handler().get_current_state(
+                room.room_id
+            )
```
please can you give this `event_type` and `state_key` arguments. Pulling out the entire room state for a large room is quite expensive.

jkanefendt replied:
Solved in https://github.com/matrix-org/synapse/pull/10384/commits/29664b170d3ac19c49c1c3f813187bbbb8666a6a

richvdh commented on synapse/federation/transport/server/__init__.py:
```diff
@@ -286,7 +288,50 @@ async def on_GET(
+            )
+            room_power_levels = room_state.get((EventTypes.PowerLevels, ""))
+            if room_power_levels is not None:
+                stripped_content = copy.deepcopy(room_power_levels.content)
+                stripped_content["users"] = {
+                    k: v for k, v in stripped_content["users"].items() if k == user_id
+                }
```
a comment explaining what we're doing here, with an example, would be helpful. It's not entirely clear at first glance.

richvdh reviewed (commented):
Generally this looks great to me, though I have a few nits it would be nice to clean up.

richvdh commented:
One other thing: it would be really nice to see some tests for this stuff, especially the power-levels stuff which looks fiddly. It might be best to wait for more progress on the MSC before putting in too much work in that direction though.

anoadragon453 commented:
Hi again @jkanefendt. Just checking in to see if you're able to continue with the suggested changes? It'd be good to see this PR finished up :)

jkanefendt commented on synapse/storage/databases/main/openid.py:
```diff
@@ -12,32 +12,41 @@
         self, token: str, ts_now_ms: int
-    ) -> Optional[str]:
-        def get_user_id_for_token_txn(txn: LoggingTransaction) -> Optional[str]:
+    ) -> Optional[Tuple[str, Optional[List[str]]]]:
```
https://github.com/matrix-org/synapse/pull/10384/commits/48a9bf1aba11a8de26d235c50ba3593431c4384a

jkanefendt replied:
https://github.com/matrix-org/synapse/pull/10384/commits/ed7f4b2b9394557ca79b4be2e4d3592f25d339c1

jkanefendt commented on synapse/federation/federation_server.py:
```diff
@@ -972,9 +972,13 @@ async def on_get_missing_events(
-    async def on_openid_userinfo(self, token: str) -> Optional[str]:
+    async def on_openid_userinfo(
+        self, token: str
+    ) -> Optional[Tuple[str, Optional[List[str]]]]:
```
https://github.com/matrix-org/synapse/pull/10384/commits/48a9bf1aba11a8de26d235c50ba3593431c4384a

jkanefendt commented on synapse/storage/schema/main/delta/70/01openid_token_userinfo_fields.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE open_id_tokens ADD COLUMN userinfo_fields TEXT;
```
https://github.com/matrix-org/synapse/pull/10384/commits/e145f9bd87afc0925aff0b9c4894fbffa38cf769

anoadragon453 commented:
Hi @jkanefendt, thank you for updating this PR.

Looking at [MSC3356](https://github.com/matrix-org/matrix-spec-proposals/pull/3356), there have been [concerns raised](https://github.com/matrix-org/matrix-spec-proposals/pull/3356#pullrequestreview-942757833) about whether the design is the right approach. I'm a little hesitant to merge this PR before that opinion in engaged, even behind an experimental flag. Could you respond with your thoughts?

DMRobertson commented:
Taking this out of the review queue until the discussions on the MSC are resolved.
